src/main/scala/com/fulcrumgenomics/bam/FindTechnicalReads.scala

```diff
@@ -77,33 +93,66 @@ class FindTechnicalReads
       val r1 = iter.next()
       val r2 = if (iter.hasNext && iter.head.getReadName == r1.getReadName) Some(iter.next()) else None
       val recs = Seq(Some(r1), r2).flatten
-      if (recs.exists(r => matcher.matches(r.getReadBases))) {
-        recs.foreach(out.addAlignment)
-        n += recs.size
+
+      val isTechnical = if (isUnmappedTemplate(r1)) {
+        val hit      = recs.toStream.flatMap(r => matcher.findMatch(r.getReadBases)).headOption
+        val hitIndex = hit.map(h => sequenceToIndex(h))
+
+        for (t <- tag; hi <- hitIndex; r <- recs) {
+          r.setAttribute(t, hi)
+          n += 1
+        }
+
+        hit.isDefined
+      }
+      else {
+        false
       }
 
+      if (isTechnical || allReads) recs.foreach(out.addAlignment)
       recs.foreach(progress.record)
     }
 
-    logger.info(s"Wrote out ${n} reads that matched to technical sequences.")
+    logger.info(s"Found ${n} reads that matched to technical sequences.")
     in.safelyClose()
     out.close()
   }
+
+  /** Returns true if all reads from the template are unmapped, otherwise false. */
+  private[bam] def isUnmappedTemplate(rec: SAMRecord) : Boolean = rec.getReadUnmappedFlag && (!rec.getReadPairedFlag || rec.getMateUnmappedFlag)
 }
 
 /** Little utility class to match reads to technical sequences. */
 private[bam] class Matcher(val matchLength: Int, val maxErrors: Int, sequences: Seq[String]) {
-  private val kmers: Array[Array[Byte]] =
-    sequences.map(_.toUpperCase).flatMap(a => Seq(a, SequenceUtil.reverseComplement(a)))
+  // Map of kmer to first sequence in which it is found
+  private val kmerToSequence: mutable.LinkedHashMap[String,String] = mutable.LinkedHashMap[String,String]()
+  sequences.foreach(s => {
+    Seq(s.toUpperCase).flatMap(a => Seq(a, SequenceUtil.reverseComplement(a)))
       .flatMap(a => a.sliding(matchLength))
       .filterNot(_.contains('N'))
-      .map(_.getBytes)
-      .toSet.toArray
+      .foreach(k => if (!kmerToSequence.contains(k)) kmerToSequence(k) = s)
+  })
+
+  private val kmers: Array[Array[Byte]] = kmerToSequence.keySet.map(_.getBytes).toArray
 
+  /** Just tests to see if any sequence matches. */
   def matches(read: Array[Byte]): Boolean = {
     read.length >= matchLength && kmers.exists(kmer => matches(read=read, kmer=kmer, len=matchLength))
   }
 
+  /** Identifies the first sequence with a matching kmer. */
+  def findMatch(read: Array[Byte]): Option[String] = {
+    if (read.length < matchLength) {
+      None
+    }
+    else {
+      kmers.find(kmer => matches(read=read, kmer=kmer, len=matchLength)) match {
```

Comment:
for `findMatchIndex` could use `kmers.zipWithIndex.find { case (kmer, i) => matches(...) } .map (_._2) ...